substratevm/src/com.oracle.svm.configure/src/com/oracle/svm/configure/config/ConfigurationType.java

```diff
@@ -316,7 +316,6 @@ public void setAllPublicConstructors() {
 
     @Override
     public void printJson(JsonWriter writer) throws IOException {
```

Comment:
`ConfigurationType` should really have a `ConfigurationPredicate` field, and `ConfigurationPredicate` should implement `JsonPrintable`.

Reply:
`ConfigurationPredicate` can't implement `JsonPrintable` as it is in the SDK module. It is there as it is used by `org.graalvm.nativeimage.hosted.RuntimeReflection` and could potentially become part of the API. This was moved here to avoid the transition of the JSON machinery to the SDK.

> ConfigurationType should really have a ConfigurationPredicate field
Why? I find it easier to keep it in the map and look it up by `ConfigurationPredicate` and `name`. We need this functionality anyhow so moving `ConfigurationType` here would be extra code IMO.

Reply:
Yes, we should still have it in the map keys. But with your changes, each configuration type is predicated, explicitly or implicitly, and `JsonPrintable.printJson` is supposed to be self-contained and the whole situation in `TypeConfiguration.printJson` and `.printPredicateJson` looks unclean. Since `ConfigurationPredicate` can't implement `JsonPrintable`, we need some kind of adapter/proxy/wrapper class.

Reply:
This will yield more complexity (one extra invariant to check) and more code to satisfy the invariant (to me unknown) that `printJson` should start with `{` and end with `}`. This will also be happening in every config file we have.

To me adding extra complexity for the same functionality usually means we are diverging from a minimal model. I think the underlying question is whether predicates belong to `TypeConfiguration` or `ConfigurationType` or both. Currently, I only see the real need for them in `TypeConfiguration`. Is there any other benefit to adding this complexity than the `printJson` integrity?

Reply:
`ConfigurationType` is the direct counterpart of a type object in the JSON file (i.e. what you also refer to as a configuration entry), and with the JSON notation you introduced, the predicate is part of a type object, so it is natural for the predicate to be an attribute of `ConfigurationType`, and for `ConfigurationType.printJson` to print the entire object's representation and not a sequence of fields which the caller needs to know how to handle. The storage in a map in `TypeConfiguration` in which the predicate is a part of the key can be seen as an optimization, in the same way that the qualified name of the type is an attribute of `ConfigurationType`, but was used as key in `TypeConfiguration` before your changes.

Reply:
Tried it out and it seems to reduce overall complexity.

Reply:
Fixed.